File: typescript/vfs.ts
import * as fs from 'fs';
import * as path from 'path';

export class VirtualFileSystem {
    private files: Map<string, Buffer> = new Map();

    public writeFile(filePath: string, content: Buffer): void {
        this.files.set(filePath, content);
    }

    public async readFile(filePath: string): Promise<Buffer> {
        const content = this.files.get(filePath);
        if (content) {
            return content;
        }
        return fs.promises.readFile(filePath);
    }

    public async flush(): Promise<void> {
        for (const [filePath, content] of this.files.entries()) {
            await fs.promises.writeFile(filePath, content);
            console.log(`[hermes:vfs] Flushed: ${filePath}`);
        }
        this.files.clear();
    }
}

File: csharp/Vfs.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Hermes
{
    public class VirtualFileSystem
    {
        private readonly Dictionary<string, byte[]> _files = new Dictionary<string, byte[]>();
        private readonly object _lock = new object();

        public void WriteFile(string path, byte[] content)
        {
            lock (_lock)
            {
                _files[path] = content;
            }
        }

        public byte[] ReadFile(string path)
        {
            lock (_lock)
            {
                if (_files.TryGetValue(path, out var content))
                {
                    return content;
                }
            }

            return File.ReadAllBytes(path);
        }

        public void Flush()
        {
            lock (_lock)
            {
                foreach (var kvp in _files)
                {
                    File.WriteAllBytes(kvp.Key, kvp.Value);
                    Console.WriteLine($"[hermes:vfs] Flushed: {kvp.Key}");
                }
                _files.Clear();
            }
        }
    }
}
